/* @brief Implements serialization for measurements on the Soil Power Sensor
 * 
 * Each individual measurement has all identifying information for someone to
 * determine when the measurement happened, who performed the measurement, what
 * was being measured, and the measurement itself. In the scheme of the Soil
 * Power Sensor, single measurements are taken, then saved to NVRAM for a later
 * upload. This was perferred over using "repeated" types as the data would have
 * to be uploaded in blocks of LoRaWAN due to payload size limitations.
 * 
 * @author John Madden <jmadden173@pm.me>
 * @date 2023-11-21
 */

syntax = "proto3";

/* Data shared between all measurement messages */
message MeasurementMetadata {
  // id of the cell measured
  uint32 cell_id = 1;
  // id of the logging device
  uint32 logger_id = 2;
  // timestamp of the measurement
  uint32 ts = 3;
}

/* Power measurement message. Voltage and current can be digitially combined to
 * obtain power. */
message PowerMeasurement {
  // voltage
  double voltage = 2;
  // current
  double current = 3;
}

/* Teros12 measurement message */
message Teros12Measurement {
  // raw volumetric water content
  double vwc_raw = 2;
  // calibrated volumetric water content
  double vwc_adj = 3;
  // temperature in celcious
  double temp = 4;
  // electrical conductivity
  uint32 ec = 5;
}

/* Phytos measurement */
message Phytos31Measurement {
  // raw adc voltage
  double voltage = 1;
  // calibrated leaf wetness
  double leaf_wetness = 2;
}

/* Top level measurement message */
message Measurement {
  // Metadata
  MeasurementMetadata meta = 1;

  // Possible measurements
  oneof measurement {
    PowerMeasurement power = 2;
    Teros12Measurement teros12 = 3;
    Phytos31Measurement phytos31 = 4;
  }
}

/* Acknowledge Packet */
message Response {
  /* Response codes from server */
  enum ResponseType {
    /* Data was successfully uploaded */
    SUCCESS = 0;
    /* General Error */
    ERROR = 1;
  }

  /* Response from server */
  ResponseType resp = 1; 
}

message Esp32Command {
  oneof command {
    PageCommand page_command = 1;
    TestCommand test_command = 2;
    WiFiCommand wifi_command = 3;
  }
}

message PageCommand {
  /* Type of request */
  enum RequestType {
    OPEN = 0;
    CLOSE = 1;
    READ = 2;
    WRITE = 3;
  }

  /* File request type */
  RequestType file_request = 1;
  /* Integer for file descriptor */
  uint32 file_descriptor = 2;
  /* Block size of read/write */
  uint32 block_size = 3;
  /* Number of bytes */
  uint32 num_bytes = 4;
}

message TestCommand {
  enum ChangeState {
    /* Data is received by the module */
    RECEIVE = 0;
    /* Data is received by the module indicating data for subsequent request */
    RECEIVE_REQUEST = 1;
    /* Data is sent my the module */
    REQUEST = 2;
  }

  /* State to put module into */
  ChangeState state = 1;
  /* Data field for test command data */
  int32 data = 2;
}

message WiFiCommand {
  enum Type {
    /* Connect to WiFi network*/
    CONNECT = 0;
    /* Post data to endpoint */
    POST = 1;
  }

  /* Command type */
  Type type = 1;

  /* Group for Connect */
  /* WiFI SSID */
  string ssid = 2;
  /* WiFi Password */
  string passwd = 3;
  /* Endpoint url */
  string url = 4;
  /* Port */
  uint32 port = 8;
 
  /* Group for request repsonses */
  /* Return code */
  uint32 rc = 5;
  /* Timestamp n unix epochs */
  uint32 ts = 6;
  // binary data response
  bytes resp = 7;
}